server/schedule/checker/rule_checker.go

```diff
@@ -135,22 +131,64 @@ func (c *RuleChecker) CheckWithFit(region *core.RegionInfo, fit *placement.Regio
 	return nil
 }
 
-func (c *RuleChecker) fixRulePeer(region *core.RegionInfo, fit *placement.RegionFit, rf *placement.RuleFit) (*operator.Operator, error) {
+func (c *RuleChecker) fixRulePeer(region *core.RegionInfo, fit *placement.RegionFit, rf *placement.RuleFit, skipOrphan bool) (*operator.Operator, error) {
 	// make up peers.
 	if len(rf.Peers) < rf.Rule.Count {
 		return c.addRulePeer(region, rf)
 	}
-	// fix down/offline peers.
+
+	if !skipOrphan && !rf.IsSatisfied() {
+		checkerCounter.WithLabelValues("rule_checker", "skip-remove-orphan-peer").Inc()
+		skipOrphan = true
+	}
+
+	var (
+		selectedPeer *metapb.Peer
+		kind         string
+	)
 	for _, peer := range rf.Peers {
-		if c.isDownPeer(region, peer) {
-			checkerCounter.WithLabelValues("rule_checker", "replace-down").Inc()
-			return c.replaceUnexpectRulePeer(region, rf, fit, peer, downStatus)
+		// fix down/offline peers.
+		if selectedPeer == nil && c.isDownPeer(region, peer) {
+			selectedPeer, kind = peer, downStatus
+		}
+		if selectedPeer == nil && c.isOfflinePeer(peer) {
+			selectedPeer, kind = peer, offlineStatus
 		}
-		if c.isOfflinePeer(peer) {
-			checkerCounter.WithLabelValues("rule_checker", "replace-offline").Inc()
-			return c.replaceUnexpectRulePeer(region, rf, fit, peer, offlineStatus)
+
+		// check orphan peer.
+		if !skipOrphan {
+			for _, pendingPeer := range region.GetPendingPeers() {
```

Comment:
how about define some function like  containsPeer(peers,peerID) bool?
if it is only used in this func, define it as a temp variable should be ok, for example,
```
	constainsPeer := func(peers []*pdpb.PeerStats, peerID uint64) bool {
		for _, peer := range peers {
			if peer.Peer.Id == peerID {
				return true
			}
		}
		return false
	}
```

Reply:
The down peer and pending peer are stored in the different struct, which cannot share the same function.